### DISCUSSION TITLE
Compilation error with Additional Module "Support for SD/MMC card via hardware/software SPI"

### DISCUSSION THREAD

**Discussion by mvackel** [Paciente8159/uCNC#506] [Config and issues] [answered]
Using uCNC 1.7.4, Mega/RAMPS 1.4 with ReprapDiscount full display smart contoller.

I'm trying to add the SD card to the offline controller, but the compilation fails with the following warning & error:

uCNC\src\modules\SD_card\sd_card.c: In function 'sd_card_file_print':
uCNC\src\modules\SD_card\sd_card.c:175:4: warning: implicit declaration of function 'serial_flush'; did you mean 'serial_putc'? [-Wimplicit-function-declaration]
    serial_flush();
    ^~~~~~~~~~~Compiling .pio\build\ramps14\src\src\modules\softuart.c.o
~
    serial_putc

Linking .pio\build\ramps14\firmware.elf
C:\Users\Mack\AppData\Local\Temp\ccTV82ie.ltrans0.ltrans.o: In function `settings_save':
<artificial>:(.text+0xa0d8): undefined reference to `serial_flush'

(What are the "Additional modules" "Enable extension options" that should be enabled to compile for SD card? I enabled them all....   Maybe it's a good idea to enable the necessary options as the user chooses a module).

**Comment by Paciente8159** [comment]
I'll check.

> (What are the "Additional modules" "Enable extension options" that should be enabled to compile for SD card? I enabled them all.... Maybe it's a good idea to enable the necessary options as the user chooses a module).

Yes that as crossed my mind. I just haven't found the time to implement that. It's on the bucket list

**Comment by Paciente8159** [comment]
> Linking .pio\build\ramps14\firmware.elf
> C:\Users\Mack\AppData\Local\Temp\ccTV82ie.ltrans0.ltrans.o: In function `settings_save': <artificial>:(.text+0xa0d8): undefined reference to `serial_flush'

That function is deprecated and is no longer used. Comment that line.
I will update the modules. Also note that the SD card PF version will replace the non PF.

I will also hide that module option from download and add the deprecated warning

**Reply by Paciente8159** [reply, marked as answer]
Updates applied.

**Reply by mvackel** [reply]
The SD card module compiled ok and seems to be working - although the interface if very simplified. I'll post the results of a real test as soon as I got the lathe running.

A related question: should the SD_CARD_PF be used instead of the SD_CARD module? What are the differences/advantages?

**Reply by Paciente8159** [reply]
Mainly how it's coded internally. It allows to seamless use Petit FS instead of FatFs and also make better use of some features and settings of FatFS.

PetitFS although very limited as a tiny footprint. My plan was to add it a couple of funtions in the future to make ir more usefull. For MCU with limited Flash (even some 32 bit ones) this might come in handy.

**Reply by mvackel** [reply]
The SD_CARD_PF module seems to be causing some issues in the UGS. The error messages are:

[MSG:SD card mounted]
[MSG:SD card settings not found]
An error was detected while sending '$G': (error:7) An EEPROM read failed. Auto-restoring affected EEPROM to default values. Streaming has been paused.


[GC:G1 G17 G90 G94 G21 G49 G54 G61 M5 M9 M48 T1 F0.000 S0]
Error while processing response <ok>: An unexpected command was completed by the controller.

**Reply by Paciente8159** [reply]
Thanks. I will investigate.
